### PR TITLE
Add {IoSlice, IoSliceMut}::advance

### LINKED ISSUE #62726: Tracking issue for io_slice_advance
This is a tracking issue for `IoSlice::{advance, advance_slices}` and `IoSliceMut::{advance, advance_slices}`.

Feature gate: #![feature(io_slice_advance)].

Steps:

- [x] Implement the RFC.
    * #62987
    * #85802
- [ ] FCP
- [ ] Stabilization PR.

Current API additions:

```rust
impl<'a> IoSlice<'a> { // And `IoSliceMut`
    pub fn advance(&mut self, n: usize);
    pub fn advance_slices(bufs: &mut &mut [IoSlice<'a>], n: usize);
}
```

---
Old issue:

Writing https://github.com/rust-lang-nursery/futures-rs/pull/1741 I needed to resort to unsafe code to change the underlying slice in `IoSlice` (and `IoSliceMut`). I'm missing a method that can change the underlying slice. @Nemo157 said that I should open an issue.

Current idea would be something like `Buf::advance` from the [bytes crate](https://docs.rs/bytes/0.4.12/bytes/trait.Buf.html#tymethod.advance).

```rust
impl IoSlice {
    // Advance the internal cursors of the slice by `n` bytes.
    fn advance(&mut self, n: usize) {
        // ..
    }
}
```

### PR DESCRIPTION
**Pull request by Thomasdezeeuw**
API inspired by the [`Buf::advance`](https://docs.rs/bytes/0.4.12/bytes/trait.Buf.html#tymethod.advance) method found in the [bytes](https://docs.rs/bytes) crate.

Closes #62726.

### CODE REVIEW

File: src/libstd/sys/cloudabi/io.rs [hunk 1]
```diff
@@ -20,6 +27,13 @@ impl<'a> IoSliceMut<'a> {
         IoSliceMut(buf)
     }
 
+    #[inline]
+    pub unsafe fn advance(&mut self, n: usize) {
+        // We create a new slice unsafely to get around lifetime issues. Since
+        // the slice need to life at least as long lifetime 'a, this is safe.
+        self.0 = slice::from_raw_parts_mut(self.0.as_mut_ptr().add(n), self.0.len() - n)
```
Comment (Mark-Simulacrum) [inline src/libstd/sys/cloudabi/io.rs]: Why is this necessary? It should be possible to "move" the slice along without unsafe code, perhaps [`split_at_mut`](https://doc.rust-lang.org/nightly/std/primitive.slice.html) would be of help here?

Comment (Thomasdezeeuw) [reply src/libstd/sys/cloudabi/io.rs]: As the comment says it's to work around lifetime issue, see this [playground example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ff0c3cc39ff05c315a511911ff58a4b9).

**Comment by alexcrichton** [comment]
Thanks for the PR! As a convenience function though I would imagine that wouldn't this be more useful if defined on something like `&[IoVec]` or `&mut [IoVecMut]`? Specifically you typically have a list of buffers and you want to advance them all forward rather than just one at a time which still has its own trickiness in management.

**Comment by Thomasdezeeuw** [comment]
> As a convenience function though I would imagine that wouldn't this be more useful if defined on something like `&[IoVec]` or `&mut [IoVecMut]`?

I agree, that is what I need for https://github.com/rust-lang-nursery/futures-rs/pull/1741 as well. I'm fine with changing it. But where would I put the code for such a change? In the `slice` file or in `IoSlice` file?

**Comment by alexcrichton** [comment]
Perhaps something like:

```rust
impl IoVec<'_> {
    pub fn advance(slice: &mut [IoVec]) { /* ... */ }
}

impl IoVecMut<'_> {
    pub fn advance(slice: &mut [IoVecMut]) { /* ... */ }
}
```

or something like that?

**Comment by Thomasdezeeuw** [comment]
Looks good, I'll try to update the pr tomorrow.

File: src/libstd/io/mod.rs [hunk 2]
```diff
@@ -942,6 +942,55 @@ impl<'a> IoSliceMut<'a> {
...
+        for buf in bufs.iter() {
+            if accumulated_len + buf.len() > n {
+                break;
+            } else {
+                accumulated_len += buf.len();
+                remove += 1;
+            }
+        }
+
+        *bufs = unsafe {
```
Comment (alexcrichton) [inline src/libstd/io/mod.rs]: I think it'd be good to remove the `unsafe` here since this is unlikely to be performance critical, and the "idiomatic" way to typically do this looks like:

```rust
use std::mem;

pub fn advance<T>(buf: &mut &mut [T], amt: usize) {
    let slice = mem::replace(buf, &mut []);
    *buf = &mut slice[amt..];
}
```

File: src/libstd/io/mod.rs [hunk 3]
```diff
@@ -942,6 +942,55 @@ impl<'a> IoSliceMut<'a> {
...
+    /// ][..];
+    ///
+    /// // Mark 10 bytes as read.
+    /// IoSliceMut::advance(&mut bufs, 10);
+    /// assert_eq!(bufs[0].deref(), [2; 14].as_ref());
+    /// assert_eq!(bufs[1].deref(), [3; 8].as_ref());
+    /// ```
+    #[unstable(feature = "io_slice_advance", issue = "62726")]
+    #[inline]
+    pub fn advance(bufs: &mut &mut [IoSliceMut<'a>], n: usize) {
```
Comment (alexcrichton) [inline src/libstd/io/mod.rs]: One other option for an API signature is:

```rust
pub fn advance(bufs: &'b mut [IoSliceMut<'a>], n: usize) -> &'b mut [IoSliceMut<'a>]
```

but that's perhaps a bit wonky since it still mutates the contents themselves. It's also somewhat wonky though to take `&mut &mut`, so I'm not really sure if there's a slam dunk one way or another.

File: src/libstd/io/mod.rs [hunk 4]
```diff
@@ -989,6 +1038,54 @@ impl<'a> IoSlice<'a> {
...
+                remove += 1;
+            }
+        }
+
+        *bufs = unsafe {
+            slice::from_raw_parts_mut(bufs.as_mut_ptr().add(remove), bufs.len() - remove)
+        };
+
+        if !bufs.is_empty() {
+            unsafe { bufs[0].0.advance(n - accumulated_len) }
```
Comment (alexcrichton) [inline src/libstd/io/mod.rs]: I may have missed the previous discussion, but how come these internal `advance` methods are `unsafe`?

File: src/libstd/io/mod.rs (hunk 3 above)
Comment (alexcrichton) [reply src/libstd/io/mod.rs]: Hm can you elaborate why it wouldn't work? I'm not sure I understand what the limitation would be

File: src/libstd/io/mod.rs (hunk 4 above)
Comment (alexcrichton) [reply src/libstd/io/mod.rs]: Is the `unsafe` here then because there's no bounds checking performed on the amount passed in? Is it necessary to have `unsafe` here for performance/code size/etc? (I suspect we don't need this `unsafe` for now and we'll be fine with having the implementation be safe which is why I'm asking)

File: src/libstd/sys/cloudabi/io.rs (hunk 1 above)
Comment (alexcrichton) [reply src/libstd/sys/cloudabi/io.rs]: FWIW I personally agree with @Mark-Simulacrum here, this (and the other impls) should be doable without unsafe code

File: src/libstd/io/mod.rs (hunk 3 above)
Comment (Thomasdezeeuw) [reply src/libstd/io/mod.rs]: For some reason I didn't think that API would work when storing `&mut [IoSliceMut]` in a struct, but it does.

I do think your suggested API is nicer, however I do have one reservation wouldn't it be surprising that the original `IoSliceMut` is sometimes modified? People might expect that effectively a sub slice of the original `bufs` is returned even though they're passing in `&mut`. If you don't think its a problem I'll change the API.

Comment (alexcrichton) [reply src/libstd/io/mod.rs]: While I see them both as a bit wonky, I think I prefer the API that returns a slice with a doc comment indicating that the elements may be mutated.

Comment (Thomasdezeeuw) [reply src/libstd/io/mod.rs]: I've updated the API.

File: src/libstd/io/mod.rs [hunk 5]
```diff
@@ -942,6 +942,57 @@ impl<'a> IoSliceMut<'a> {
     pub fn new(buf: &'a mut [u8]) -> IoSliceMut<'a> {
         IoSliceMut(sys::io::IoSliceMut::new(buf))
     }
+
+    /// Advance the internal cursor of the slice.
+    ///
+    /// # Notes
+    ///
+    /// Elements in the slice may be modified.
```
Comment (Thomasdezeeuw) [inline src/libstd/io/mod.rs]: This might need expanding.

**Comment by Thomasdezeeuw** [comment]
I've addressed all comments and updated the API to return the truncated slice.

File: src/libstd/io/mod.rs [hunk 6]
```diff
@@ -989,6 +1040,56 @@ impl<'a> IoSlice<'a> {
...
         IoSlice(sys::io::IoSlice::new(buf))
     }
+
+    /// Advance the internal cursor of the slice.
+    ///
+    /// # Notes
+    ///
+    /// Elements in the slice may be modified.
+    ///
+    /// # Examples
```
Comment (alexcrichton) [inline src/libstd/io/mod.rs]: The documentation here I think may also want to mention that this API panics if you advance beyond the cumulative size of all buffers.

File: src/libstd/io/mod.rs (hunk 5 above)
Comment (alexcrichton) [reply src/libstd/io/mod.rs]: Yeah I'd recommend going into detail here and explaining how the vector elements of the slice may be modified to adjust the size as noted by `advance`

File: src/libstd/io/mod.rs (hunk 6 above)
Comment (Thomasdezeeuw) [reply src/libstd/io/mod.rs]: But it doesn't panic if you go beyond the cumulative size, see the `io_slice_advance_beyond_total_length` test.

File: src/libstd/io/mod.rs [hunk 7]
```diff
@@ -942,6 +942,58 @@ impl<'a> IoSliceMut<'a> {
...
+    /// let mut buf2 = [2; 16];
+    /// let mut buf3 = [3; 8];
+    /// let mut bufs = &mut [
+    ///     IoSliceMut::new(&mut buf1),
+    ///     IoSliceMut::new(&mut buf2),
+    ///     IoSliceMut::new(&mut buf3),
+    /// ][..];
+    ///
+    /// // Mark 10 bytes as read.
+    /// bufs = IoSliceMut::advance(mem::replace(&mut bufs, &mut []), 10);
```
Comment (Thomasdezeeuw) [inline src/libstd/io/mod.rs]: I'm not too fond of this `mem::replace` dance we have to do.

File: src/libstd/io/mod.rs (hunk 5 above)
Comment (Thomasdezeeuw) [reply src/libstd/io/mod.rs]: What do you think of what I've added?

**Comment by Thomasdezeeuw** [comment]
The commits have become a bit messy, I'll squash them before merging once the review is complete.

**Comment by alexcrichton** [comment]
Ok everything looks great to me, thanks! If you want to do some cleanup feel free!

@bors: delegate+

**Comment by Thomasdezeeuw** [comment]
This is ready now, I've squashed the commits and rebased on master.

**Comment by Thomasdezeeuw** [comment]
Hope this works.

@bors r+

File: src/libstd/sys/windows/io.rs [hunk 8]
```diff
@@ -48,6 +60,18 @@ impl<'a> IoSliceMut<'a> {
         }
     }
 
+    #[inline]
+    pub fn advance(&mut self, n: usize) {
+        if (self.vec.len as usize) < n {
+            panic!("advancing IoSliceMut beyond its length");
+        }
+
+        unsafe {
```
Comment (Centril) [inline src/libstd/sys/windows/io.rs]: Could we always uphold the minimum standard of having comments on all `unsafe {` blocks?
I think that's something you should be able to expect of something so central as the standard library.
